### PR TITLE
Implement Firestore IndexBackfiller

### PR DESCRIPTION
**Pull request by tom-andersen**
Implement index backfiller

### CODE REVIEW

File: packages/firestore/src/local/local_documents_view.ts [hunk 1]
```diff
@@ -56,6 +57,13 @@ import { PersistencePromise } from './persistence_promise';
 import { PersistenceTransaction } from './persistence_transaction';
 import { RemoteDocumentCache } from './remote_document_cache';
 
+export class LocalDocumentsResult {
```
Comment (wu-hui) [inline packages/firestore/src/local/local_documents_view.ts:60]: Comment.

File: packages/firestore/src/local/index_manager.ts [hunk 2]
```diff
@@ -138,4 +138,18 @@ export interface IndexManager {
...
+  /**
+   * Iterates over all field indexes that are used to serve the given target,
+   * and returns the minimum offset of them all.
+   */
+  getMinOffset(
+    transaction: PersistenceTransaction,
+    target: Target
+  ): PersistencePromise<IndexOffset>;
+
+  getMinOffsetFromCollectionGroup(
```
Comment (ehsannas) [inline packages/firestore/src/local/index_manager.ts:151]: Add comment

File: packages/firestore/src/local/persistence_promise.ts [hunk 3]
```diff
@@ -243,4 +243,31 @@ export class PersistencePromise<T> {
...
+   * return `true`, the process repeats, until `condition` returns false.
+   * @param condition
+   * @param action
+   */
+  static loopUntil(
+    condition: () => boolean,
+    action: () => PersistencePromise<void>
+  ): PersistencePromise<void> {
+    return new PersistencePromise<void>((resolve, reject) => {
+      const process = (): void => {
```
Comment (ehsannas) [inline packages/firestore/src/local/persistence_promise.ts:261]: I think the preferred style is to remove `:void`
```js
const process = () => {
  // ...
}
```

**Review by ehsannas — COMMENTED** [review]
I didn't realize that this was assigned to Andy until I had read a few of the files. So, it's just a drive-by review, not a thorough one. Please refer to Andy for approval.

File: packages/firestore/src/local/lru_garbage_collector_impl.ts [hunk 4]
```diff
@@ -110,18 +110,18 @@ class RollingSequenceNumberBuffer {
...
-export class LruScheduler implements GarbageCollectionScheduler {
-  private hasRun: boolean = false;
+export class LruScheduler implements Scheduler {
   private gcTask: DelayedOperation<void> | null;
 
   constructor(
     private readonly garbageCollector: LruGarbageCollector,
-    private readonly asyncQueue: AsyncQueue
+    private readonly asyncQueue: AsyncQueue,
+    private readonly localStore: LocalStore
```
Comment (ehsannas) [inline packages/firestore/src/local/lru_garbage_collector_impl.ts:119]: I'm also curious why LocalStore became a member variable. Is it diverging from the other SDKs?

File: packages/firestore/src/local/local_documents_view.ts (hunk 1 above)
Comment (ehsannas) [reply packages/firestore/src/local/local_documents_view.ts:60]: So the Web SDK has an interface called `LocalWriteResult`. [Ref](https://cs.opensource.google/firebase-sdk/firebase-js-sdk/+/master:packages/firestore/src/local/local_store_impl.ts;l=100-103?q=localwriteresult&ss=firebase-sdk), which is the same thing I believe.

The iOS SDK has a similar thing. [Ref](https://cs.opensource.google/firebase-sdk/firebase-ios-sdk/+/master:Firestore/core/src/local/local_write_result.h;l=32?q=localwriteresult&ss=firebase-sdk).

The Web SDK so far has not used `LocalDocumentResult` so far, and the tests also use `LocalWriteResult`.

I think we should use `LocalWriteResult` in this PR.

Potentially, _if we think LocalWriteResult is the better way to go_, we should make a separate PR that swaps all usecases of `LocalWriteResult` with `LocalDocumentResult`. Alternatively, we can update Android to match Web and iOS.

thoughts?

Comment (ehsannas) [reply packages/firestore/src/local/local_documents_view.ts:60]: either way we must make sure all of them use the same name

**Comment by ehsannas** [comment]
@tom-andersen There are a number of tests in `SQLiteLocalStoreTest.java` that are missing in the Web SDK. They should also be ported as a part of this PR. I think the reason they haven't been ported so far was that some of them rely on the `backfillIndexes()` API which is being added in this PR.

File: repo-scripts/prune-dts/.run/AllTests.run.xml [hunk 5]
```diff
@@ -2,7 +2,7 @@
   <configuration default="false" name="All Tests" type="mocha-javascript-test-runner">
     <node-interpreter>project</node-interpreter>
     <node-options />
-    <mocha-package>$PROJECT_DIR$/../../node_modules/mocha</mocha-package>
```
[the remark this answers is not in the document]
Comment (tom-andersen) [reply repo-scripts/prune-dts/.run/AllTests.run.xml]: Reverted

File: packages/firestore/src/local/local_documents_view.ts (hunk 1 above)
Comment (tom-andersen) [reply packages/firestore/src/local/local_documents_view.ts]: Changed to `LocalWriteResult`

File: packages/firestore/src/local/local_documents_view.ts [hunk 6]
```diff
@@ -319,6 +327,61 @@ export class LocalDocumentsView {
...
+                count - originalDocs.size
+              )
+            : PersistencePromise.resolve(newOverlayMap());
+        let largestBatchId = INITIAL_LARGEST_BATCH_ID;
+        let modifiedDocs = originalDocs;
+        return overlaysPromise.next(overlays => {
+          return PersistencePromise.forEach(
+            overlays,
+            (key: DocumentKey, overlay: Overlay) => {
+              if (largestBatchId < overlay.largestBatchId) {
```
[the remark this answers is not in the document]
Comment (tom-andersen) [reply packages/firestore/src/local/local_documents_view.ts:400]: Comment copied

File: packages/firestore/src/local/persistence_promise.ts (hunk 3 above)
Comment (ehsannas) [reply packages/firestore/src/local/persistence_promise.ts:285]: huh! TIL

File: packages/firestore/src/local/indexeddb_index_manager.ts [hunk 7]
```diff
@@ -1062,3 +1031,23 @@ function indexStateStore(
...
+
+function getMinOffsetFromFieldIndexes(fieldIndexes: FieldIndex[]): IndexOffset {
+  hardAssert(
+    fieldIndexes.length !== 0,
+    'Found empty index group when looking for least recent index offset.'
+  );
+
+  let minOffset: IndexOffset = fieldIndexes[0].indexState.offset;
+  let maxBatchId: number = minOffset.largestBatchId;
+  for (let i = 1; i < fieldIndexes.length; i++) {
```
Comment (ehsannas) [inline packages/firestore/src/local/indexeddb_index_manager.ts:1043]: I'm curious why you switched from `for (const fieldIndex of fieldIndexes) {` ?
I _think_ that's the preferred way in JS if you don't really care about the loop index value

**Review by ehsannas — COMMENTED** [review]
you can run `yarn prettier && yarn lint:fix` (in `packages/firestore` directory) before you run `git push` so you can catch format/lint issues locally.

**Review by ehsannas — COMMENTED** [review]
Thanks for addressing my comments. Pass to Andy for thorough review and approval.

File: packages/firestore/src/local/indexeddb_index_manager.ts [hunk 8]
```diff
@@ -951,6 +969,56 @@ export class IndexedDbIndexManager implements IndexManager {
...
+      minOffset.readTime,
+      minOffset.documentKey,
+      maxBatchId
+    );
+  }
+
+  getMinOffset(
+    transaction: PersistenceTransaction,
+    target: Target
+  ): PersistencePromise<IndexOffset> {
```
[the remark this answers is not in the document]
Comment (tom-andersen) [reply packages/firestore/src/local/indexeddb_index_manager.ts:981]: Did option #1

File: packages/firestore/src/local/indexeddb_index_manager.ts (hunk 7 above)
Comment (ehsannas) [reply packages/firestore/src/local/indexeddb_index_manager.ts:1043]: but now you're assuming the array has at least 2 elements, right? `fieldIndexes[1]` shouldn't be accessed if we are trying to find the minOffset of an array of size 1.

File: packages/firestore/src/local/index_backfiller.ts [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+        await this.schedule(REGULAR_BACKFILL_DELAY_MS);
+      }
+    );
+  }
+}
+
+/** Implements the steps for backfilling indexes. */
+export class IndexBackfiller {
+  constructor(
+    private readonly localStore: LocalStore,
```
Comment (wu-hui) [inline packages/firestore/src/local/index_backfiller.ts:112]: Add a comment to explain why we need localstore here.

File: packages/firestore/src/local/local_store.ts [hunk 10]
```diff
@@ -18,11 +18,22 @@
...
 import { logDebug } from '../util/log';
 
+import { IndexManager } from './index_manager';
+import { LocalDocumentsView } from './local_documents_view';
 import { LruGarbageCollector, LruResults } from './lru_garbage_collector';
 import { PRIMARY_LEASE_LOST_ERROR_MSG } from './persistence_transaction';
 
 export interface LocalStore {
   collectGarbage(garbageCollector: LruGarbageCollector): Promise<LruResults>;
+
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply packages/firestore/src/local/local_store.ts:28]: As discussed offline, this change makes user change better to handle, even though it derails the convention. We will go forward with this, and change Android to match this.

File: packages/firestore/src/local/index_backfiller.ts (hunk 9 above)
Comment (tom-andersen) [reply packages/firestore/src/local/index_backfiller.ts:119]: Comment added

File: packages/firestore/src/model/collections.ts [hunk 11]
```diff
@@ -54,20 +55,24 @@ export function documentMap(...docs: Document[]): DocumentMap {
   return map;
 }
 
-export type OverlayMap = ObjectMap<DocumentKey, Overlay>;
-export function newOverlayMap(): OverlayMap {
-  return new ObjectMap<DocumentKey, Overlay>(
-    key => key.toString(),
-    (l, r) => l.isEqual(r)
+export function convertDocumentKeyMapWithOverlayedDocumentToDocumentMap(
```
[the remark this answers is not in the document]
Comment (tom-andersen) [reply packages/firestore/src/model/collections.ts]: Introduced type OverlayedDocumentMap (following existing pattern), and renamed method accordingly.

File: packages/firestore/src/local/indexeddb_index_manager.ts [hunk 12]
```diff
@@ -951,26 +966,26 @@ export class IndexedDbIndexManager implements IndexManager {
...
-            indexOffsetComparator(index.indexState.offset, offset) < 0
-          ) {
-            offset = index.indexState.offset;
-          }
-        });
-      }
-    ).next(() => offset!);
+      (subTarget: Target) =>
+        this.getFieldIndex(transaction, subTarget).next(index =>
+          !index ? fail('Target cannot be served from index') : index
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply packages/firestore/src/local/indexeddb_index_manager.ts]: Sounds good.

**Pull request merged**